avantifellows/plio-frontend
# Get new access tokens when expired, auto logout used if refresh token is expired/deleted, encrypt localStorage data (tokens)

Fixes #210 
Fixes #190 
Fixes #219 

## Summary
- refreshes access token if it is expired given the refresh token has not expired
- logs the user out if refresh token has expired
- encrypts token in localStorage 



- Added a library `secure-ls` to encrypt while storing and decrypt while retrieving store data that's being saved to localStorage by `vuex-persistedstate`. More details [here](https://github.com/robinvdvleuten/vuex-persistedstate#encrypted-local-storage)
- Changed `VUE_APP_BACKEND_AUTH_URL` in `.env`. 
- Added two endpoints to `Endpoints.js`

   ```js
   convertTokenEndpoint = "/convert-token/"
   refreshTokenEndpoint = "/token/"
   ```
- This is done such that we get a base URL of ```http://0.0.0.0:8001/auth``` and add different auth endpoints on top of that. The two auth endpoints that we will use will be
    ```js
    VUE_APP_BACKEND_AUTH_URL + convertTokenEndpoint // for gauth token conversion
    VUE_APP_BACKEND_AUTH_URL + refreshTokenEndpoint // for getting a new access token
    ```
- For the refresh token logic, instead of implementing a new view function at the backend, I have just used the functionality offered by the **Social OAuth2** package that we're using for google auth. This saves us a custom implementation where we would've have to verify the refresh token, generate a new access token, revoke the old refresh token and other steps. How to use a refresh token to retrieve a new access token using **Social Oauth2** can be seen on the readme [here](https://github.com/RealmTeam/django-rest-framework-social-oauth2#testing-the-setup).
- To test in Postman - 
    ```js
    POST "http://0.0.0.0:8001/auth/token/"
    // request
    {
        "grant_type": "refresh_token",
        "client_id": "...",
        "client_secret": "...",
        "refresh_token": "2pmJHJCmHnGLSRlexJdQag1qA2i6Sj"
    }

    // response
    {
        "access_token": "WVLUSQkF4i51yNIWh4RKA1sK99QuSP",
        "expires_in": 86400,
        "token_type": "Bearer",
        "scope": "read write",
        "refresh_token": "HvOatAiMzL2hS5pnKbd2MO3XMc5NLA"
    }
    ```

- Added a new vuex store state variable - "loginStatus" (true or false). We'll watch for this variable in `App.vue` and if true, we'll log out the user and show them a dialog
- Added a dialog box to `App.vue` to show the popup "You've been logged out. Please log in again"
- The whole flow for handling 401 un authorized calls is inspired from [this](https://stackoverflow.com/questions/51563821/axios-interceptors-retry-original-request-and-access-original-promise) with some minor tweaks
- The case of handling 400 requests on the refreshTokenEndpoint itself is not mentioned in that post and I've implemented that on my own
- To test expiration of access_token, go to TablePlus and reduce the expiry date of the latest issued `access_token` and try to navigate on the platform. Everything should work fine
- To test expiration/revoked refresh_token, go to TablePlus, do the above step first, then go to the refresh_token table, make the entry in `access_token_id` column as `NULL` and enter some datetime in the column `revoked`. Basically we're emulating the revokation of a refresh token. Its important to emulate the expiry of the `access_token` along with this, otherwise the client will keep using the access_token for all requests. Expected behaviour - you will be logged out! 


## Test Plan
Tested locally.
